### ISSUE TITLE
Allow synchronious execution of signal process function

### ISSUE THREAD

**Issue by xfactor2000** [godotengine/godot#19136]
**Godot version:**
Godot Mono 3.0.2 official

**OS/device including version:**
Manjaro Linux 17.1.10

**Issue description:**
While writing my implementation of eventbus in Godot, I discovered that the signal process function is called asynchroniously. This is not good, especially if the same signal arrives several times without delay. Godot fires the function in parallel each time a signal arrives, and it means the state of the Node may change before the first signal processing has the chance to complete.

I tried Mutex, it didn't help because both functions run on the same thread.

Proposal - allow for synchronious processing of signals, like in Akka Actors:
https://stackoverflow.com/questions/7772764/actors-scala-akka-is-it-implied-that-the-receive-method-will-be-accessed-in-a

This could solve a lot of potential issues with using signals. Right now, I'm forced to use C# delegates.

**Comment by Zylann** [comment]
What is the "signal process function"?

**Comment by xfactor2000** [comment]
The function I use in Node.Connect as callback

**Comment by Zylann** [comment]
But, aren't these synchronous already? I mean, when you `emit_signal`, all handlers get called immediately, there is no asynchronous handling unless you specify it with `CONNECT_DEFERRED`.

**Comment by neikeq** [comment]
> But, aren't these synchronous already? I mean, when you `emit_signal`, all handlers get called immediately, there is no asynchronous handling unless you specify it with `CONNECT_DEFERRED`.

You are correct:

https://github.com/godotengine/godot/blob/aee8d262528497298a50ff2b25eefd8dce18c033/core/object.cpp#L1191-L1195

Maybe OP misunderstood something?

**Comment by xfactor2000** [comment]
No, what I meant is when inside the handler and **another** signal arrives , then the handler is fired again without waiting for the processing of the first signal to finish. I guess it happens in another thread, although when I print System.Threading.Thread.CurrentManagedThreadId, it  always stays the same.

**Comment by Zylann** [comment]
> I guess it happens in another thread

It doesn't. Handlers get fired immediately because by default, they are synchronous, that's what it is. If you want them to wait for the current handler to finish, maybe you should use CONNECT_DEFERRED instead, which queues them. But it's possible that it won't get queued the way you expect if the engine is already processing that queue in the first place.

**Comment by xfactor2000** [comment]
Can you please elaborate how do they fire in parallel if no additional thread is used? I can't quite get my head around it.

**Comment by xfactor2000** [comment]
> That means your handler fired the signal it was processing, which makes it called a second time before the first completed? (recursion?) As far as I know, there is no thread involved unless you specifically ask for it.

OK, I can get around that, but what if another signal arrives in my handler while it's processing, and I'm not talking about signal it fired? Will that signal wait until my handler finishes?

**Comment by Zylann** [comment]
@xfactor2000 if a signal gets emitted a second time while you are already handling it, and if the SAME handler gets called, it can only be due to your handler triggering it itself by some sort of recursion. If that's not the case, then some other code might be doing a nasty usage of threads.

**Comment by xfactor2000** [comment]
OK, now I understand. And you're saying that using connect deferred flag application-wise is not a good idea?

**Comment by Zylann** [comment]
Connect deferred schedules a call on the main thread to be dequeued on the next idle frame AFAIK (which makes it suitable for threads to emit signals to be handled in the main thread).

**Comment by xfactor2000** [comment]
OK, 
guess I'll have to find out.I will nudge the people in power to add this to documentation.

Thanks a ton for all the help,
case closed (or is it?)
